File: Topological_Sorting/line_up.cpp
// A -> B : A가 B보다 앞에 있어야함.
#include <bits/stdc++.h>
using namespace std;

const int MX = 32000;
vector<int> adj[MX];
int indeg[MX];
int n, m;

void toplogical(){
    queue<int> Q;
    vector<int> res;

    for (int i=1; i<=n; i++){
        if (indeg[i]==0) Q.push(i);
    }
    while (!Q.empty()){
        int cur = Q.front(); Q.pop();
        res.push_back(cur);

        for (int nxt : adj[cur]){
            indeg[nxt] --;
            if (indeg[nxt] == 0) Q.push(nxt);
        }
    }
    if (res.size() != n)
        cout << "Graph has Cycle : NO DAG";
    else {
        for (auto x : res) cout << x << " ";
    }
}

int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> n >> m;  // node, edge
    while(m--){
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
        indeg[b]++;
    }
    toplogical();

    return 0;
}

File: Topological_Sorting/line_up.py
from collections import deque 

MX = 1005
adj = [[] for _ in range(MX)]
indeg = [0] * MX 
n, m = map(int, input().split())

def topological():
    Q = deque()
    res = []
    
    for i in range(1, n+1):
        if (indeg[i] == 0):
            Q.append(i)
    
    while (Q):
        cur = Q.popleft()
        res.append(cur)
        
        for nxt in adj[cur]:
            indeg[nxt] -= 1
            if (indeg[nxt] == 0):
                Q.append(nxt)
    
    if (len(res) != n):
        print("No DAG. Graph has cycle")
    
    else:
        for x in res:
            print(x, end = " ")

for _ in range(m):
    a, b = map(int, input().split())
    adj[a].append(b)
    indeg[b] += 1
    
topological()
